### ISSUE TITLE
Apparent formatter speed regression

### ISSUE THREAD

**Issue by mattbaker** [elixir-lsp/elixir-ls#381]
Editing a simple `.exs` test file (340 lines) is taking around 3 seconds to format on save with no other edited files. I get a loading bar and a notification at bottom right mentioning the `"ElixirLS: Elixir support and debugger"` formatter running.

I installed 0.5.0 manually from the vsix and the problem is gone, so it looks like a regression in the latest release. 

Any guesses on what happened between versions? Any particular logs or other diagnostic information can I provide to help debug?

This is an umbrella project of a fair size. Pure Elixir, so no large node_modules directories, etc.

Possible relation to #378?

### Environment

* Elixir & Erlang versions (elixir --version): Elixir 1.9.4 (compiled with Erlang/OTP 22)
* Operating system: MacOS Mojave
* Editor or IDE name (e.g. Emacs/VSCode): VSCode
* Editor Plugin/LSP Client name: elixir-lsp/elixir-ls

### Troubleshooting

- [x] Restart your editor (which will restart ElixirLS) sometimes fixes issues
- [x] Stop your editor, remove the entire `.elixir_ls` directory, then restart your editor
  * NOTE: This will cause you to have to re-run the dialyzer build for your project

### Logs

Wasn't sure exactly what to provide in this case but happy to include what I can.

**Comment by BlueHotDog** [comment]
Same here :(

**Comment by axelson** [comment]
Do you have a sample project that exhibits this issue? And what is in your `.formatter.exs` inputs?

**Comment by mattbaker** [comment]
Unfortunately I don't have a sample repro :/ 

We're in an umbrella project, so here's the contents of our various formatter configs:

```
> fd --hidden ".formatter.exs" --exec cat {}
[
  inputs: ["{mix,.formatter}.exs"],
  subdirectories: ["apps/*"]
]

[
  inputs: ["{mix,.formatter}.exs", "{config,lib,test}/**/*.{ex,exs}"]
]

[
  inputs: ["{mix,.formatter}.exs", "{config,lib,test}/**/*.{ex,exs}"]
]

[
  inputs: ["{mix,.formatter}.exs", "{config,lib,test}/**/*.{ex,exs}"]
]

[
  inputs: ["{mix,.formatter}.exs", "{config,lib,test}/**/*.{ex,exs}"]
]

[
  inputs: ["{mix,.formatter}.exs", "{config,lib,test}/**/*.{ex,exs}"]
]

[
  inputs: ["{mix,.formatter}.exs", "{config,lib,test}/**/*.{ex,exs}"]
]

[
  inputs: ["{mix,.formatter}.exs", "{config,lib,test}/**/*.{ex,exs}"]
]

[
  inputs: [
    "{lib,config,test}/**/*.{ex,exs}",
    "{mix,.formatter}.exs"
  ],
  import_deps: [:absinthe, :plug],
  locals_without_parens: [
    forward: 1,
    visibility: 1,
    owning_team: 1,
    experimental_complexity: 1,
    import_autostitched_types: 2,
    import_autostitched_fields: 2
  ]
]

[
  inputs: ["{mix,.formatter}.exs", "{config,lib,test}/**/*.{ex,exs}"]
]

[
  inputs: [
    "{lib,config,test}/**/*.{ex,exs}",
    "{mix,.formatter}.exs"
  ],
  import_deps: [:absinthe],
  locals_without_parens: [
    forward: 1,
    visibility: 1,
    import_autostitched_types: 2,
    import_autostitched_fields: 2
  ]
]

[
  inputs: ["{mix,.formatter}.exs", "{config,lib,test}/**/*.{ex,exs}"]
]

[
  inputs: ["{mix,.formatter}.exs", "{config,lib,test}/**/*.{ex,exs}"]
]

[
  inputs: [
    "{lib,config,test}/**/*.{ex,exs}",
    "{mix,.formatter}.exs"
  ],
  import_deps: [:plug]
]

[
  inputs: ["{mix,.formatter}.exs", "{config,lib,test}/**/*.{ex,exs}"]
]

[
  inputs: ["{mix,.formatter}.exs", "{config,lib,test}/**/*.{ex,exs}"]
]

[
  inputs: ["{mix,.formatter}.exs", "{config,lib,test}/**/*.{ex,exs}"]
]
```

**Comment by BlueHotDog** [comment]
I'm getting this in the output:
```
[Trace - 1:44:38 PM] Sending request 'textDocument/formatting - (9569)'.
[Trace - 1:44:38 PM] Received response 'textDocument/formatting - (9569)' in 1ms. Request failed: Cannot format file from current directory (Currently in deps/telemetry_poller/deps/parse_trans) (-32603).
[Error - 1:44:38 PM] Request textDocument/formatting failed.
  Message: Cannot format file from current directory (Currently in deps/telemetry_poller/deps/parse_trans)
  Code: -32603 
[Trace - 1:44:38 PM] Sending notification 'textDocument/didSave'.
[Trace - 1:44:39 PM] Sending notification 'workspace/didChangeWatchedFiles'.
[Trace - 1:44:39 PM] Sending request 'textDocument/hover - (9570)'.
[Trace - 1:44:39 PM] Sending notification '$/cancelRequest'.
[Trace - 1:44:39 PM] Received response 'textDocument/hover - (9570)' in 34ms. Request failed: Request cancelled (-32800).
```
Maybe it helps.
It resolves itself for a while if i restart the editor

**Comment by lukaszsamson** [comment]
@BlueHotDog do you have any file open from `deps/telemetry_poller/deps/parse_trans`? Also please post your `.formatter.exs`

**Comment by BlueHotDog** [comment]
this is my .formatter:
```
[
  import_deps: [:ecto, :phoenix],
  inputs: ["*.{ex,exs}", "priv/*/seeds.exs", "{config,lib,test}/**/*.{ex,exs}"],
  subdirectories: ["priv/*/migrations"]
]
```
I do not have any file from deps/delemetry.. open :(

**Comment by lukaszsamson** [comment]
Hmm the code in https://github.com/elixir-lsp/elixir-ls/blob/d7208fe42ad0a28c72d9acf2236d83cd7f785071/apps/language_server/lib/language_server/providers/formatting.ex#L43 seems strange. @axelson do you have an idea why is it checkin cwd? It was added in https://github.com/elixir-lsp/elixir-ls/commit/d1437729c0a905cad31f371eac22d4ade1195193#diff-d695db38f04b7e4e345c6f1341e4b0fd9ec92ab81ae4741577dba5ad9e7d5778 but I'm not convinced by the reasoning in the commit message.

**Comment by axelson** [comment]
@lukaszsamson I am agreed with you. I don't see a reason we can't format regardless of the CWD. Since we have the project dir we can always use that to find the relevant `.formatter.exs`. But with that said, this is a different issue that is already being tracked as #252

**Comment by benwilson512** [comment]
Can confirm that this is what we have experienced as well. Perhaps @axelson you could try using the absinthe-graphql/absinthe code base as an example? It's got quite a few files and test files.

Does elixir-ls compile the full code base before formatting? Does format on save format all files or just the one saved?

**Comment by mattbaker** [comment]
I can repro more specifically by repeatedly changing and saving a file. Eventually it'll block with the notification that the file save action is waiting on Elixir LS to format*. 

In the Elixir LS logs I noticed this seemed to occur if I saved right after I saw this message:

```
 [ElixirLS WorkspaceSymbols] Updating index...
```

During that time format/save doesn't complete until it finishes. It seems like this may correlate with saving _while_ the workspace symbol update is in progress?

If I make changes _really quickly_ it seems like I can reproduce this in 0.5.0 too. I'm unclear why it's worse in 0.6.0 though. I noticed 0.6.0 bumps Elixir and Erlang, some change there? I didn't see anything obvious in the release notes of either of those though. Is indexing slower, and thus the problem more apparent?

One way or another it seems odd that reindexing the workspace symbols is blocking the format operation. It seems like we'd want to avoid blocking a save operation at all costs. 

Is this something the Language Server spec imposes? Is something in indexing causing the formatter to have to wait? 

I'm not sure when I'll next have a chance to look into this deeply, but hopefully this helps.

----------------

\* If I go _really_ fast I get a different error that feels like another issue to address at a later date:

```
MIX_TARGET: 
[Error - 1:55:12 PM] Request textDocument/formatting failed.
  Message: Cannot format file from current directory (Currently in apps/thrift_services)
  Code: -32603 
```

**Comment by mattbaker** [comment]
@axelson - Should I close this or would you like it to remain open? I don't want to mess with your workflow.

**Comment by axelson** [comment]
I want to do some more testing and collect more feedback before closing this. Also ideally we could change the code to not use `Path.wildcard` to remove more directory traversal.

**Comment by mattbaker** [comment]
Nice, good call 👍

**Comment by jonleighton** [comment]
I've proposed a partial fix for this problem in https://github.com/elixir-lsp/elixir-ls/pull/609 (by also fixing a bug with how `:inputs` are matched). It delivers a big speed-up on a project I work on, but doesn't fix the underlying problem of relying on `Path.wildcard/2`.

I plan to also propose a more general solution, but that will take a bit longer. In the mean time, please try out the changes above and let me know if you encounter any problems (other than performance) on your projects.

**Comment by axelson** [comment]
https://github.com/elixir-lsp/elixir-ls/pull/609 was expanded to fix this issue completely since with that change, ElixirLS no longer requires a call to `Path.wildcard/2` for formatting now that [PathGlob](https://github.com/jonleighton/path_glob) is used instead :confetti_ball:

**Comment by mattbaker** [comment]
That’s amazing! 🎉

On December 4, 2021, GitHub ***@***.***> wrote:
> #609 <https://github.com/elixir-lsp/elixir-ls/pull/609> was expanded
> to fix this issue completely since with that change, ElixirLS no
> longer requires a call to Path.wildcard/2 for formatting now that
> PathGlob <https://github.com/jonleighton/path_glob> is used instead 🎊
>

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/elixir-lsp/elixir-ls/issues/381#issuecomment-
986160421>, or unsubscribe
<https://github.com/notifications/unsubscribe-
auth/AAANRJSMRTYEYQB23ERA4W3UPLM75ANCNFSM4SL6QKNQ>.
